Update BMW connected drive to async

rikroe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
- Removed `binary_sensor.<your_vehicle>_lights_parking` (not provided by API anymore)
- Renamed sensors. The existing sensors with historic data and automations should be migrated automatically.
  - `sensor.<your_vehicle>_charging_level_hv` to `sensor.<your_vehicle>_remaining_battery_percent`
  - `sensor.<your_vehicle>_fuel_percent` to `sensor.<your_vehicle>_remaining_fuel_percent`

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update `bimmer_connected` library to `0.9.0` which has been completely reworked and is now async-based with `httpx`.
Except for the breaking changes above, everything is still as before.

Fixes:
- login for the `china` region, which was broken due to being more obscured by BMW.
- fixes #71559.

The code (and the library) have been tested via HACS since 2022-04-10 using our [custom component](https://github.com/bimmerconnected/ha_custom_component/releases/tag/20220410.2).

~~I'm opening this as draft as the component is currently using the latest beta release of the component. I'm waiting for another contributor but wanted to get some feedback if I missed something on the component itself or need to change something.~~

Release notes: https://github.com/bimmerconnected/bimmer_connected/releases/tag/0.9.0
Diff: https://github.com/bimmerconnected/bimmer_connected/compare/0.8.12..0.9.0

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

rikroe committed:
Change BMW connected drive to async

rikroe commented:
Hey @MartinHjelmare, could you have a look at this even in draft?
As I couldn't answer to https://github.com/home-assistant/core/pull/67003#pullrequestreview-965294213 anymore, I included the changes in this PR.

- Remove `Entity` from `BMWConnectedDriveBaseEntity(CoordinatorEntity, Entity)`
- The `isinstance` check is not needed anymore as the account is now created in the coordinator init without directly making external HTTP calls
- Regarding `_handle_coordinator_update` I'm still not really clear whats the preferred way. I actually like the entity properties more (looks cleaner). Should I revert back to properties or keep it like it is?

Thanks for your time!

rikroe committed:
Fix coordinator exceptions, fix tests

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/config_flow.py:
```diff
@@ -32,21 +33,21 @@ async def validate_input(
     try:
-        await hass.async_add_executor_job(
-            ConnectedDriveAccount,
+        account = MyBMWAccount(
```
Please only wrap the line that can raise in the try... except block. If this doesn't do I/O we can move it up.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/coordinator.py:
```diff
@@ -20,52 +24,44 @@
     async def _async_update_data(self) -> None:
         """Fetch data from BMW."""
         try:
+            old_refresh_token = self.account.refresh_token
```
As above.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/coordinator.py:
```diff
@@ -20,52 +24,44 @@
-                    )
-        except OSError as err:
+                await self.account.get_vehicles()
+            if self.account.refresh_token != old_refresh_token:
```
Please move this down.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/lock.py:
```diff
@@ -55,27 +55,27 @@ def __init__(
             self._attr_is_locked = True
-            self.schedule_update_ha_state()
-        self.vehicle.remote_services.trigger_remote_door_lock()
+            self.async_schedule_update_ha_state()
```
```suggestion
            self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/lock.py:
```diff
@@ -55,27 +55,27 @@ def __init__(
             self._attr_is_locked = False
-            self.schedule_update_ha_state()
-        self.vehicle.remote_services.trigger_remote_door_unlock()
+            self.async_schedule_update_ha_state()
```
```suggestion
            self.async_write_ha_state()
```

MartinHjelmare commented:
Whether or not to overwrite `_handle_coordinator_update` needs to be decided per integration. In some cases it's efficient to overwrite it while in some cases its better to use the entity properties.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/binary_sensor.py:
```diff
@@ -114,29 +115,19 @@ class BMWBinarySensorEntityDescription(
         },
     ),
-    BMWBinarySensorEntityDescription(
-        key="lights_parking",
```
Is this sensor removed?

rikroe committed:
Fix using deprecated property

rikroe commented:
> Whether or not to overwrite `_handle_coordinator_update` needs to be decided per integration. In some cases it's efficient to overwrite it while in some cases it's better to use the entity properties.

Ok, but in our case there shouldn't be any difference? Both `_handle_coordinator_update` as well as the computation of the entity property happen only once every 5 minutes? Or am I missing something?

If yes (and it makes no difference here), I would rather use the entity property approach as then I don't have to worry about forgetting calling `super()._handle_coordinator_update()`.

rikroe committed:
Write HA state directly

rikroe committed:
Remove code that cannnot throw Exceptions from try/except

MartinHjelmare commented:
I didn't mean efficient as in performant, I meant efficient as in less code.

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/lock.py:
```diff
@@ -74,7 +74,7 @@ async def async_unlock(self, **kwargs: Any) -> None:
             # update callback response
             self._attr_is_locked = False
-            self.async_schedule_update_ha_state()
+            self._async_write_ha_state()
```
```suggestion
            self.async_write_ha_state()
```

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/lock.py:
```diff
@@ -63,7 +63,7 @@ async def async_lock(self, **kwargs: Any) -> None:
             # update callback response
             self._attr_is_locked = True
-            self.async_schedule_update_ha_state()
+            self._async_write_ha_state()
```
```suggestion
            self.async_write_ha_state()
```

rikroe committed:
Use public async_write_ha_state

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

rikroe committed:
Fix login using refresh_token if token expired

rikroe committed:
MyPy fixes

rikroe commented:
Added some further changes.

First things first, the library (including the released wheel) is now fully typed. I therefore have adjusted the component accordingly and removed the last typing issues.

Secondly, I wasn't able to use `hass.config_entries.async_update_entry()` and had to revert to `hass.config_entries._async_schedule_save()` (d943437afb51ea83038a894679441c0f3aa1900f). Using the first method always triggered a full reload of the component, leading to an infinite loop of updating the `refresh_token`. 
I don't want a component refresh, just to persist the most current refresh token (which can only be used once).
Hope using the private function is fine, otherwise I'm happy for suggestions.

rikroe committed:
Fix pytest, bump dependency

rikroe committed:
Replace OptionFlow listener with explicit refresh

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/config_flow.py:
```diff
@@ -104,6 +104,17 @@ async def async_step_account_options(
+            # Required as each successful login will store the latest refresh_token
+            # using async_update_entry, which would otherwise trigger a full reload
+            # if the options would be refreshed using a listener.
+            if entry := self.hass.config_entries.async_get_entry(self.handler):
```
We already have `self.config_entry`.

rikroe replied:
Oops, thank you. Took me too long to figure out how to best reload so I totally missed this!

rikroe committed:
Remove uneeded async_get_entry

rikroe commented:
On a sidenote: I hope to get the library to a stable version today or latest tomorrow (latest tests using the custom component are currently running). Would this be sufficient to include in `2022.5`?

rikroe committed:
Update test to include change on OptionsFlow

MartinHjelmare commented:
Beta cut is on Wednesday. If we merge before the cut it will make the June release.

rikroe committed:
Bump bimmer_connected to 0.9.0

rikroe commented:
Just released 0.9.0 stable and updated the PR description to the final changelog and diff.
For easier comparison, here are the diff originally mentioned in the PR and from there to stable:
Original diff: https://github.com/bimmerconnected/bimmer_connected/compare/v0.9.0b10..0.9.0
Second diff: https://github.com/bimmerconnected/bimmer_connected/compare/0.8.12..v0.9.0b10

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Can we migrate the renamed sensors to the new unique_id and keep the user customizations?

We have helpers to look up and migrate entity registry entries. Eg:
https://github.com/home-assistant/core/blob/800410ddf09dad2bab842e7fa40c59568e40ff98/homeassistant/helpers/entity_registry.py#L996-L1011

rikroe committed:
Migrate renamed entitity unique_ids

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -113,6 +140,35 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+
+    Copied and adjusted from `homeassistant.helpers.entity_registry.async_migrate_entries`
+    as migration will fail (and therefore component will not load) if the new unique_id
+    has already been created (i.e. using a custom component or dev version).
```
I don't think we should copy this whole function here. Just add a check for an existing entry with the new unique and return None in that case in the callback.

rikroe replied:
Oh, does the callback has access to the registry to check if the id exists? As far as I can tell neither a `hass` object nor an `EntityRegistry` is available there?

MartinHjelmare replied:
Since we define the callback within `async_setup_entry` it has access to `hass` and any other name in the outer scope.

rikroe replied:
Ah, thanks! Learned something today :) I'll implement it accordingly.

rikroe commented:
I've implemented the migration. However I had to copy and adjust `async_migrate_entries`: As quite a lot of people are already using our custom component (which has the state of this PR), their entities have not been migrated. If such user now tries update to 2022.6, the update (and therefore component startup) will fail completely as the new unique_ids already exist.

Therefore I'm checking against those and delete the old entities (so that they won't be renamed).

If you prefer any other sort of error handling (like try ... except around `async_migrate_entries` and just don't change anything on `ValueError`) I can integrate those.

MartinHjelmare reviewed (commented):
Please also add a test for the migration. Create a couple of existing entries in the entity registry, set up the integration and assert that the entries are migrated ok.

rikroe committed:
Don't replace async_migrate_entries, add tests

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -91,12 +95,31 @@ def update_unique_id(entry: RegistryEntry) -> dict[str, str] | None:
                 entry.unique_id,
                 new_unique_id,
             )
+            if existing_entry := entity_registry.async_get_entity_id(
```
```suggestion
            if existing_entity_id := entity_registry.async_get_entity_id(
```

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -91,12 +95,31 @@ def update_unique_id(entry: RegistryEntry) -> dict[str, str] | None:
+                _LOGGER.debug(
+                    "Cannot migrate to unique_id '%s', already exists for '%s'",
+                    new_unique_id,
+                    existing_entry,
```
```suggestion
                    existing_entity_id,
```

MartinHjelmare commented on homeassistant/components/bmw_connected_drive/__init__.py:
```diff
@@ -1,31 +1,32 @@
 import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.entity import DeviceInfo, Entity
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_registry import (
```
If we import the entity registry module as `er`, `async_get` will be a bit clearer what it is when we use it with `er.async_get`.

MartinHjelmare commented:
Here's an example of a test to test the migration:
https://github.com/home-assistant/core/blob/070cb616317f4661e8558706ae025d3d9b2b4868/tests/components/plugwise/test_init.py#L73-L111

rikroe committed:
Rename existing_entry to existing_entry_id

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

rikroe committed:
Update tests

rikroe committed:
Import full EntityRegistry

MartinHjelmare commented on tests/components/bmw_connected_drive/test_init.py:
```diff
@@ -1,116 +1,136 @@
-    registry.async_get_or_create(
+    entity_registry = er.async_get(hass)
+
+    # create existin gentry with new_unique_id
```
```suggestion
    # create existing entry with new_unique_id
```

MartinHjelmare committed:
Fix comment

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Please update the breaking change paragraph and mention the migration and what users can expect, ie if they need to do anything or not.

rikroe committed:
Increase timeout to 60s

MartinHjelmare commented:
Maybe remove the mention of hardcoded YAML? Since we migrate the unique_id of the entities where we rename entity_id, the entity_id will stay the same for existing users.

rikroe commented:
I removed the YAML part and changed the changes to *should*, indicating that everything should be migrated but they still have to watch out.

rikroe committed:
Rely on library timeout

rikroe commented:
Done. We're getting there I hope :)

Btw, again thanks for all your time and patience, always a pleasure!

MartinHjelmare commented:
🎉
